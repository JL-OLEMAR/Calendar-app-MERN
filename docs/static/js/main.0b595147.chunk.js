(this["webpackJsonpcalendar-app"]=this["webpackJsonpcalendar-app"]||[]).push([[0],{115:function(e,t,n){},117:function(e,t,n){"use strict";n.r(t);var a=n(8),r=n(9),c=n(0),o=n(36),s=n(10),i=n(19),u=n(12),l=n.n(u),d=n(18),j=n.n(d),b="http://localhost:4000/api",m=function(e,t){var n=arguments.length>2&&void 0!==arguments[2]?arguments[2]:"GET",a="".concat(b,"/").concat(e);return"GET"===n?window.fetch(a):window.fetch(a,{method:n,headers:{"Content-Type":"application/json"},body:JSON.stringify(t)})},f=function(e,t){var n=arguments.length>2&&void 0!==arguments[2]?arguments[2]:"GET",a="".concat(b,"/").concat(e),r=window.localStorage.getItem("token")||"";return"GET"===n?window.fetch(a,{method:n,headers:{"x-token":r}}):window.fetch(a,{method:n,headers:{"Content-Type":"application/json","x-token":r},body:JSON.stringify(t)})},p="[ui] Open modal",O="[ui] Close modal",v="[event] Set active",h="[event] Logout event",x="[event] Add new",g="[event] Clear active event",w="[event] Event updaed",y="[event] Event deleted",N="[event] Events loaded",k="[auth] Finish checking login state",E="[auth] Login",S="[auth] Logout",C=n(5),T=n(17),D=n.n(T),P=function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:[];return e.map((function(e){return Object(C.a)(Object(C.a)({},e),{},{start:D()(e.start).toDate(),end:D()(e.end).toDate()})}))},I=function(e){return{type:x,payload:e}},A=function(){return{type:g}},H=function(e){return{type:w,payload:e}},_=function(){return{type:y}},L=function(e){return{type:N,payload:e}},M=function(){return{type:k}},R=function(e){return{type:E,payload:e}},G=function(){return{type:S}},F=n(15),V=n(27),J=function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:{},t=Object(c.useState)(e),n=Object(F.a)(t,2),a=n[0],r=n[1],o=function(){r(e)},s=function(e){var t=e.target;r(Object(C.a)(Object(C.a)({},a),{},Object(V.a)({},t.name,t.value)))};return[a,s,o]},q=(n(74),n(2)),U=function(){var e=Object(r.b)(),t=J({lEmail:"",lPassword:""}),n=Object(F.a)(t,2),a=n[0],c=n[1],o=J({rName:"",rEmail:"",rPassword1:"",rPassword2:""}),s=Object(F.a)(o,2),u=s[0],d=s[1],b=a.lEmail,f=a.lPassword,p=u.rName,O=u.rEmail,v=u.rPassword1,h=u.rPassword2;return Object(q.jsx)("div",{className:"container login-container",children:Object(q.jsxs)("div",{className:"row",children:[Object(q.jsxs)("div",{className:"col-md-6 login-form-1",children:[Object(q.jsx)("h3",{children:"Ingreso"}),Object(q.jsxs)("form",{onSubmit:function(t){var n,a;t.preventDefault(),e((n=b,a=f,function(){var e=Object(i.a)(l.a.mark((function e(t){var r,c;return l.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return e.next=2,m("auth",{email:n,password:a},"POST");case 2:return r=e.sent,e.next=5,r.json();case 5:(c=e.sent).ok?(window.localStorage.setItem("token",c.token),window.localStorage.setItem("token-init-date",(new Date).getTime()),t(R({uid:c.uid,name:c.name}))):j.a.fire("Error",c.msg,"error");case 7:case"end":return e.stop()}}),e)})));return function(t){return e.apply(this,arguments)}}()))},children:[Object(q.jsx)("div",{className:"form-group",children:Object(q.jsx)("input",{className:"form-control",name:"lEmail",placeholder:"Correo",type:"text",value:b,onChange:c})}),Object(q.jsx)("div",{className:"form-group",children:Object(q.jsx)("input",{className:"form-control",name:"lPassword",placeholder:"Contrase\xf1a",type:"password",value:f,onChange:c})}),Object(q.jsx)("div",{className:"form-group",children:Object(q.jsx)("input",{className:"btnSubmit",type:"submit",value:"Login"})})]})]}),Object(q.jsxs)("div",{className:"col-md-6 login-form-2",children:[Object(q.jsx)("h3",{children:"Registro"}),Object(q.jsxs)("form",{onSubmit:function(t){if(t.preventDefault(),v!==h)return j.a.fire("error","Las contrase\xf1as deben de ser iguales","error");var n,a,r;e((n=O,a=v,r=p,function(){var e=Object(i.a)(l.a.mark((function e(t){var c,o;return l.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return e.next=2,m("auth/new",{email:n,password:a,name:r},"POST");case 2:return c=e.sent,e.next=5,c.json();case 5:(o=e.sent).ok?(window.localStorage.setItem("token",o.token),window.localStorage.setItem("token-init-date",(new Date).getTime()),t(R({uid:o.uid,name:o.name}))):j.a.fire("Error",o.msg,"error");case 7:case"end":return e.stop()}}),e)})));return function(t){return e.apply(this,arguments)}}()))},children:[Object(q.jsx)("div",{className:"form-group",children:Object(q.jsx)("input",{className:"form-control",name:"rName",placeholder:"Nombre",type:"text",value:p,onChange:d})}),Object(q.jsx)("div",{className:"form-group",children:Object(q.jsx)("input",{className:"form-control",name:"rEmail",placeholder:"Correo",type:"email",value:O,onChange:d})}),Object(q.jsx)("div",{className:"form-group",children:Object(q.jsx)("input",{className:"form-control",name:"rPassword1",placeholder:"Contrase\xf1a",type:"password",value:v,onChange:d})}),Object(q.jsx)("div",{className:"form-group",children:Object(q.jsx)("input",{className:"form-control",name:"rPassword2",placeholder:"Repita la contrase\xf1a",type:"password",value:h,onChange:d})}),Object(q.jsx)("div",{className:"form-group",children:Object(q.jsx)("input",{className:"btnSubmit",type:"submit",value:"Crear cuenta"})})]})]})]})})},B=n(54),X=(n(78),n(79),function(){return{type:p}}),z={allDay:"Todo el d\xeda",previous:"<",next:">",today:"Hoy",month:"Mes",week:"Semana",day:"D\xeda",agenda:"Agenda",date:"Fecha",time:"Hora",event:"Evento",noEventsInRange:"No hay eventos en este rango",showMore:function(e){return"+ Ver m\xe1s (".concat(e,")")}},K=function(){var e=Object(r.b)();return Object(q.jsx)("button",{className:"btn btn-primary fab",onClick:function(){e(X())},children:Object(q.jsx)("i",{className:"fas fa-plus"})})},Q=function(){var e=Object(r.b)();return Object(q.jsxs)("button",{className:"btn btn-danger fab-danger",onClick:function(){e(function(){var e=Object(i.a)(l.a.mark((function e(t,n){var a,r,c;return l.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return a=n().calendar.activeEvent.id,e.prev=1,e.next=4,f("events/".concat(a),{},"DELETE");case 4:return r=e.sent,e.next=7,r.json();case 7:(c=e.sent).ok?t(_()):j.a.fire("Error",c.msg,"error"),e.next=14;break;case 11:e.prev=11,e.t0=e.catch(1),console.log(e.t0);case 14:case"end":return e.stop()}}),e,null,[[1,11]])})));return function(t,n){return e.apply(this,arguments)}}())},children:[Object(q.jsx)("i",{className:"fas fa-trash"}),Object(q.jsx)("span",{children:" Borrar evento "})]})},W=function(){var e=Object(r.b)(),t=Object(r.c)((function(e){return e.auth})).name;return Object(q.jsxs)("div",{className:"navbar navbar-dark bg-dark mb-4",children:[Object(q.jsx)("span",{className:"navbar-brand",children:t}),Object(q.jsxs)("button",{className:"btn btn-outline-danger",onClick:function(){e(function(){var e=Object(i.a)(l.a.mark((function e(t){return l.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:window.localStorage.clear(),t({type:h}),t(G());case 3:case"end":return e.stop()}}),e)})));return function(t){return e.apply(this,arguments)}}())},children:[Object(q.jsx)("i",{className:"fas fa-sign-out-alt"}),Object(q.jsx)("span",{children:" Salir"})]})]})},Y=function(e){var t=e.event,n=t.title,a=t.user;return Object(q.jsxs)("div",{children:[Object(q.jsx)("strong",{children:n}),Object(q.jsxs)("em",{children:[" - ",a.name]})]})},Z=n(49),$=n.n(Z),ee=n(50),te=n.n(ee),ne={content:{top:"50%",left:"50%",right:"auto",bottom:"auto",marginRight:"-50%",transform:"translate(-50%, -50%)"}};$.a.setAppElement("#root");var ae=D()().minutes(0).seconds(0).add(1,"hours"),re=ae.clone().add(1,"hours"),ce={title:"",notes:"",start:ae.toDate(),end:re.toDate()},oe=function(){var e=Object(r.c)((function(e){return e.ui})).modalOpen,t=Object(r.c)((function(e){return e.calendar})).activeEvent,n=Object(r.b)(),a=Object(c.useState)((function(){return ae.toDate()})),o=Object(F.a)(a,2),s=o[0],u=o[1],d=Object(c.useState)((function(){return re.toDate()})),b=Object(F.a)(d,2),m=b[0],p=b[1],v=Object(c.useState)(!0),h=Object(F.a)(v,2),x=h[0],g=h[1],w=Object(c.useState)((function(){return ce})),y=Object(F.a)(w,2),N=y[0],k=y[1],E=N.notes,S=N.title,T=N.start,P=N.end;Object(c.useEffect)((function(){k(t||ce)}),[t,k]);var _=function(e){var t=e.target;k(Object(C.a)(Object(C.a)({},N),{},Object(V.a)({},t.name,t.value)))},L=function(){n({type:O}),n(A()),k(ce)};return Object(q.jsxs)($.a,{ariaHideApp:!1,className:"modal",closeTimeoutMS:200,isOpen:e,overlayClassName:"modal-fondo",style:ne,onRequestClose:L,children:[Object(q.jsxs)("h1",{children:[" ",t?"Editar evento":"Nuevo evento"," "]}),Object(q.jsx)("hr",{}),Object(q.jsxs)("form",{className:"container",onSubmit:function(e){e.preventDefault();var a,r=D()(T),c=D()(P);return r.isSameOrAfter(c)?j.a.fire("Error","La fecha fin debe de ser mayor a la fecha de inicio","error"):S.trim().length<2?g(!1):(n(t?(a=N,function(){var e=Object(i.a)(l.a.mark((function e(t){var n,r;return l.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return e.prev=0,e.next=3,f("events/".concat(a.id),a,"PUT");case 3:return n=e.sent,e.next=6,n.json();case 6:(r=e.sent).ok?t(H(a)):j.a.fire("Error",r.msg,"error"),e.next=13;break;case 10:e.prev=10,e.t0=e.catch(0),console.log(e.t0);case 13:case"end":return e.stop()}}),e,null,[[0,10]])})));return function(t){return e.apply(this,arguments)}}()):function(e){return function(){var t=Object(i.a)(l.a.mark((function t(n,a){var r,c,o,s,i;return l.a.wrap((function(t){for(;;)switch(t.prev=t.next){case 0:return r=a().auth,c=r.uid,o=r.name,t.prev=1,t.next=4,f("events",e,"POST");case 4:return s=t.sent,t.next=7,s.json();case 7:(i=t.sent).ok&&(e.id=i.evento.id,e.user={_id:c,name:o},n(I(e))),t.next=14;break;case 11:t.prev=11,t.t0=t.catch(1),console.log(t.t0);case 14:case"end":return t.stop()}}),t,null,[[1,11]])})));return function(e,n){return t.apply(this,arguments)}}()}(N)),g(!0),void L())},children:[Object(q.jsxs)("div",{className:"form-group",children:[Object(q.jsx)("label",{children:"Fecha y hora inicio"}),Object(q.jsx)(te.a,{className:"form-control",format:"dd/MM/yyyy HH:mm",name:"start",required:!0,value:s,onChange:function(e){u(e),k(Object(C.a)(Object(C.a)({},N),{},{start:e}))}})]}),Object(q.jsxs)("div",{className:"form-group",children:[Object(q.jsx)("label",{children:"Fecha y hora fin"}),Object(q.jsx)(te.a,{className:"form-control",format:"dd/MM/yyyy HH:mm",minDate:s,name:"end",required:!0,value:m,onChange:function(e){p(e),k(Object(C.a)(Object(C.a)({},N),{},{end:e}))}})]}),Object(q.jsx)("hr",{}),Object(q.jsxs)("div",{className:"form-group",children:[Object(q.jsx)("label",{children:"Titulo y notas"}),Object(q.jsx)("input",{autoComplete:"off",className:"form-control ".concat(!x&&"is-invalid"),name:"title",placeholder:"T\xedtulo del evento",type:"text",value:S,onChange:_}),Object(q.jsx)("small",{className:"form-text text-muted",id:"emailHelp",children:"Una descripci\xf3n corta"})]}),Object(q.jsxs)("div",{className:"form-group",children:[Object(q.jsx)("textarea",{className:"form-control",name:"notes",placeholder:"Notas",rows:"5",type:"text",value:E,onChange:_})," ",Object(q.jsx)("small",{className:"form-text text-muted",id:"emailHelp",children:"Informaci\xf3n adicional"})]}),Object(q.jsxs)("button",{className:"btn btn-outline-primary btn-block",type:"submit",children:[Object(q.jsx)("i",{className:"far fa-save"}),Object(q.jsx)("span",{children:" Guardar"})]})]})]})};D.a.locale("es");var se=Object(B.b)(D.a),ie=function(){var e=Object(r.b)(),t=Object(r.c)((function(e){return e.calendar})),n=t.events,a=t.activeEvent,o=Object(r.c)((function(e){return e.auth})).uid,s=Object(c.useState)(window.localStorage.getItem("lastView")||"month"),u=Object(F.a)(s,2),d=u[0],j=u[1];Object(c.useEffect)((function(){e(function(){var e=Object(i.a)(l.a.mark((function e(t){var n,a,r;return l.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return e.prev=0,e.next=3,f("events");case 3:return n=e.sent,e.next=6,n.json();case 6:a=e.sent,r=P(a.eventos),t(L(r)),e.next=14;break;case 11:e.prev=11,e.t0=e.catch(0),console.log(e.t0);case 14:case"end":return e.stop()}}),e,null,[[0,11]])})));return function(t){return e.apply(this,arguments)}}())}),[e]);return Object(q.jsxs)("div",{className:"calendar-screen",children:[Object(q.jsx)(W,{}),Object(q.jsx)(B.a,{selectable:!0,components:{event:Y},endAccessor:"end",eventPropGetter:function(e,t,n,a){return{style:{backgroundColor:o===e.user._id?"#367CF7":"#465660",borderRadius:"0px",color:"white",display:"block",opacity:.8}}},events:n,localizer:se,messages:z,startAccessor:"start",view:d,onDoubleClickEvent:function(t){e(X())},onSelectEvent:function(t){e({type:v,payload:t})},onSelectSlot:function(t){e(A())},onView:function(e){j(e),window.localStorage.setItem("lastView",e)}}),Object(q.jsx)(K,{}),a&&Object(q.jsx)(Q,{}),Object(q.jsx)(oe,{})]})},ue=n(39),le=["isAuthenticated","component"],de=function(e){var t=e.isAuthenticated,n=e.component,a=Object(ue.a)(e,le);return Object(q.jsx)(s.b,Object(C.a)(Object(C.a)({},a),{},{component:function(e){return t?Object(q.jsx)(n,Object(C.a)({},e)):Object(q.jsx)(s.a,{to:"/login"})}}))},je=["isAuthenticated","component"],be=function(e){var t=e.isAuthenticated,n=e.component,a=Object(ue.a)(e,je);return Object(q.jsx)(s.b,Object(C.a)(Object(C.a)({},a),{},{component:function(e){return t?Object(q.jsx)(s.a,{to:"/"}):Object(q.jsx)(n,Object(C.a)({},e))}}))},me=function(){var e=Object(r.b)(),t=Object(r.c)((function(e){return e.auth})),n=t.checking,a=t.uid;return Object(c.useEffect)((function(){e(function(){var e=Object(i.a)(l.a.mark((function e(t){var n,a;return l.a.wrap((function(e){for(;;)switch(e.prev=e.next){case 0:return e.next=2,f("auth/renew");case 2:return n=e.sent,e.next=5,n.json();case 5:(a=e.sent).ok?(window.localStorage.setItem("token",a.token),window.localStorage.setItem("token-init-date",(new Date).getTime()),t(R({uid:a.uid,name:a.name}))):t(M());case 7:case"end":return e.stop()}}),e)})));return function(t){return e.apply(this,arguments)}}())}),[e]),n?Object(q.jsx)("h5",{children:"Espere..."}):Object(q.jsx)(o.a,{children:Object(q.jsx)("div",{children:Object(q.jsxs)(s.d,{children:[Object(q.jsx)(be,{exact:!0,component:U,isAuthenticated:!!a,path:"/login"}),Object(q.jsx)(de,{exact:!0,component:ie,isAuthenticated:!!a,path:"/"}),Object(q.jsx)(s.a,{to:"/"})]})})})},fe=n(23),pe=n(66),Oe={checking:!0},ve=n(55),he={events:[],activeEvent:null},xe={modalOpen:!1},ge=Object(fe.c)({ui:function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:xe,t=arguments.length>1?arguments[1]:void 0;switch(t.type){case p:return Object(C.a)(Object(C.a)({},e),{},{modalOpen:!0});case O:return Object(C.a)(Object(C.a)({},e),{},{modalOpen:!1});default:return e}},calendar:function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:he,t=arguments.length>1?arguments[1]:void 0;switch(t.type){case x:return Object(C.a)(Object(C.a)({},e),{},{events:[].concat(Object(ve.a)(e.events),[t.payload])});case v:return Object(C.a)(Object(C.a)({},e),{},{activeEvent:t.payload});case g:return Object(C.a)(Object(C.a)({},e),{},{activeEvent:null});case w:return Object(C.a)(Object(C.a)({},e),{},{events:e.events.map((function(e){return e.id===t.payload.id?t.payload:e}))});case y:return Object(C.a)(Object(C.a)({},e),{},{events:e.events.filter((function(t){return t.id!==e.activeEvent.id})),activeEvent:null});case N:return Object(C.a)(Object(C.a)({},e),{},{events:Object(ve.a)(t.payload)});case h:return Object(C.a)({},he);default:return e}},auth:function(){var e=arguments.length>0&&void 0!==arguments[0]?arguments[0]:Oe,t=arguments.length>1?arguments[1]:void 0;switch(t.type){case E:return Object(C.a)(Object(C.a)(Object(C.a)({},e),t.payload),{},{checking:!1});case k:return Object(C.a)(Object(C.a)({},e),{},{checking:!1});case S:return{checking:!1};default:return e}}}),we="undefined"!==typeof window&&window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__||fe.d,ye=Object(fe.e)(ge,we(Object(fe.a)(pe.a))),Ne=function(){return Object(q.jsx)(r.a,{store:ye,children:Object(q.jsx)(me,{})})};n(115);a.render(Object(q.jsx)(Ne,{}),document.getElementById("root"))},74:function(e,t,n){}},[[117,1,2]]]);
//# sourceMappingURL=main.0b595147.chunk.js.map