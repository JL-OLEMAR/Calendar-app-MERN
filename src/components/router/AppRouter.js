import { useEffect } from 'react'
import { BrowserRouter as Router, Redirect, Switch } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'

import { startChecking } from '../../actions/auth'
import { LoginScreen } from '../auth/LoginScreen'
import { CalendarScreen } from '../calendar/CalendarScreen'

import { PrivateRoute } from './PrivateRoute'
import { PublicRoute } from './PublicRoute'

export const AppRouter = () => {
  const dispatch = useDispatch()
  const { checking, uid } = useSelector((state) => state.auth)

  useEffect(() => {
    dispatch(startChecking())
  }, [dispatch])

  if (checking) {
    return <h5>Espere...</h5>
  }

  return (
    <Router>
      <div>
        <Switch>
          <PublicRoute
            exact
            component={LoginScreen}
            isAuthenticated={!!uid}
            path='/login'
          />

          <PrivateRoute
            exact
            component={CalendarScreen}
            isAuthenticated={!!uid}
            path='/'
          />

          <Redirect to='/' />
        </Switch>
      </div>
    </Router>
  )
}
