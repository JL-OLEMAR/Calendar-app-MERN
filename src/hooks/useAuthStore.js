import { useSelector, useDispatch } from 'react-redux'

import { calendarApi } from '../api'
import { clearErrorMessage, onChecking, onLogin, onLogout } from '../store/auth'
import { onClearEvents } from '../store/calendar'

export const useAuthStore = () => {
  const { errorMessage, status, user } = useSelector((state) => state.auth)
  const dispatch = useDispatch()

  const startLogin = async ({ email, password }) => {
    dispatch(onChecking())

    try {
      const { data } = await calendarApi.post('/auth', { email, password })
      const { name, token, uid } = data

      window.localStorage.setItem('token', token)
      window.localStorage.setItem('token-init-date', new Date().getTime())
      dispatch(onLogin({ name, uid }))
    } catch (error) {
      dispatch(onLogout('Incorrect credentials'))
      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 10)
    }
  }

  const startRegister = async ({ name, email, password }) => {
    dispatch(onChecking())

    try {
      const { data } = await calendarApi.post('/auth/new', {
        name,
        email,
        password
      })
      const { name: nameData, token, uid } = data

      window.localStorage.setItem('token', token)
      window.localStorage.setItem('token-init-date', new Date().getTime())
      dispatch(onLogin({ name: nameData, uid }))
    } catch (error) {
      dispatch(onLogout(error.response.data?.msg || ''))
      setTimeout(() => {
        dispatch(clearErrorMessage())
      }, 10)
    }
  }

  const checkAuthToken = async () => {
    const token = window.localStorage.getItem('token')
    if (!token) return dispatch(onLogout())

    try {
      const { data } = await calendarApi.get('/auth/renew')
      const { name, token, uid } = data

      window.localStorage.setItem('token', token)
      window.localStorage.setItem('token-init-date', new Date().getTime())
      dispatch(onLogin({ name, uid }))
    } catch (error) {
      window.localStorage.clear()
      dispatch(onLogout())
    }
  }

  const startLogout = () => {
    window.localStorage.clear()
    dispatch(onClearEvents())
    dispatch(onLogout())
  }

  return {
    errorMessage,
    status,
    user,
    checkAuthToken,
    startLogin,
    startLogout,
    startRegister
  }
}
