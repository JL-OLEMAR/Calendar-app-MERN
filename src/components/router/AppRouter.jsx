import { useEffect } from 'react'
import { BrowserRouter as Router, Redirect, Switch } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux'

import { startChecking } from '../../actions/auth.js'
import { LoginScreen } from '../auth'
import { CalendarScreen } from '../calendar'

import { PrivateRoute } from './PrivateRoute.jsx'
import { PublicRoute } from './PublicRoute.jsx'

export const AppRouter = () => {
  const dispatch = useDispatch()
  const { checking, uid } = useSelector((state) => state.auth)

  useEffect(() => {
    dispatch(startChecking())
  }, [dispatch])

  if (checking) {
    return <h5>Espere...</h5>
  }

  return (
    <Router>
      <div>
        <Switch>
          <PublicRoute
            exact
            component={LoginScreen}
            isAuthenticated={!!uid}
            path='/login'
          />

          <PrivateRoute
            exact
            component={CalendarScreen}
            isAuthenticated={!!uid}
            path='/'
          />

          <Redirect to='/' />
        </Switch>
      </div>
    </Router>
  )
}
