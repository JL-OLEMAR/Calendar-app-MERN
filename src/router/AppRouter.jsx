import { useEffect } from 'react'
import { Navigate, Route, Routes } from 'react-router-dom'

import { useAuthStore } from '../hooks'
import { Auth } from '../auth'
import { Calendar } from '../calendar'

export function AppRouter () {
  const { checkAuthToken, status } = useAuthStore()

  useEffect(() => {
    checkAuthToken()
  }, [])

  if (status === 'checking') {
    return <h3>Loading...</h3>
  }

  return (
    <Routes>
      {status === 'not-authenticated'
        ? (
          <>
            <Route path='/auth/*' element={<Auth />} />
            <Route path='/*' element={<Navigate to='/auth/login' />} />
          </>
          )
        : (
          <>
            <Route path='/' element={<Calendar />} />
            <Route path='/*' element={<Navigate to='/' />} />
          </>
          )}
    </Routes>
  )
}
