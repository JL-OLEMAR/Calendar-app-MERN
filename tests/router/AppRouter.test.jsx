import { beforeEach, expect, describe, jest, test } from '@jest/globals'
import { render, screen } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'

import { useAuthStore } from '../../src/hooks/useAuthStore.js'
import { AppRouter } from '../../src/router/AppRouter.jsx'

jest.mock('../../src/hooks/useAuthStore.js')

// Mock whole Calendar Page with a h1 to avoid hooks
jest.mock('../../src/calendar', () => ({
  Calendar: () => <h1>CalendarPage</h1>
}))

describe('Tests in <AppRouter />', () => {
  const mockCheckAuthToken = jest.fn()

  beforeEach(() => jest.clearAllMocks())

  test('should show the loading screen and call the checkAuthToken', () => {
    useAuthStore.mockReturnValue({
      status: 'checking',
      checkAuthToken: mockCheckAuthToken
    })

    render(<AppRouter />)
    expect(screen.getByText('Loading...')).toBeTruthy()
    expect(mockCheckAuthToken).toHaveBeenCalled()
  })

  test('should show the login if not authenticated', () => {
    useAuthStore.mockReturnValue({
      status: 'not-authenticated',
      checkAuthToken: mockCheckAuthToken
    })

    const { container } = render(
      <MemoryRouter initialEntries={['/auth2/dafsk/dfjasjf']}>
        {/* Redirect /auth/login */}
        <AppRouter />
      </MemoryRouter>
    )

    expect(screen.getByText('Sign In')).toBeTruthy()
    expect(container).toMatchSnapshot()
  })

  test.only('should show the calendar if authenticated', () => {
    useAuthStore.mockReturnValue({
      status: 'authenticated',
      checkAuthToken: mockCheckAuthToken
    })

    render(
      <MemoryRouter>
        <AppRouter />
      </MemoryRouter>
    )

    expect(screen.getByText('CalendarPage')).toBeTruthy()
  })
})
