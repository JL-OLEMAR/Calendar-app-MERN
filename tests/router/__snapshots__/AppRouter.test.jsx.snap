// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <AppRouter /> should show the login if not authenticated 1`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Sign In
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="lEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="lPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Register
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="rName"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="rEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="rPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="rPassword2"
              placeholder="Repeat your password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Create account"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`Tests in <AppRouter /> should show the login if not authenticated 2`] = `
<div>
  <div
    class="container login-container"
  >
    <div
      class="row"
    >
      <div
        class="col-md-6 login-form-1"
      >
        <h3>
          Sign In
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="lEmail"
              placeholder="Email"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="lPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Login"
            />
          </div>
        </form>
      </div>
      <div
        class="col-md-6 login-form-2"
      >
        <h3>
          Register
        </h3>
        <form>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="rName"
              placeholder="Name"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="rEmail"
              placeholder="Email"
              type="email"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="rPassword"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="form-group mb-2"
          >
            <input
              class="form-control"
              name="rPassword2"
              placeholder="Repeat your password"
              type="password"
              value=""
            />
          </div>
          <div
            class="d-grid gap-2"
          >
            <input
              class="btnSubmit"
              type="submit"
              value="Create account"
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;
